lib/mongo/retryable.rb

```diff
@@ -46,18 +36,22 @@ module Retryable
     # @return [ Result ] The result of the operation.
     #
     # @since 2.1.0
-    def read_with_retry(attempt = 0, &block)
+    def read_with_retry
+      attempt = 0
       begin
-        block.call
-      rescue Error::SocketError, Error::SocketTimeoutError
-        retry_operation(&block)
+        attempt += 1
+        yield
+      rescue Error::SocketError, Error::SocketTimeoutError => e
+        raise(e) if attempt > cluster.max_read_retries
+        retry_reconnect(e)
```

Comment:
I made a comment below in the retry_reconnect method about sleeping when retrying writes and the same is true here: we weren't sleeping upon Socket errors when retrying reads before so I would hesitate to change the behavior. We were only sleeping when the cluster was sharded, instead of scanning.